From ce680f0002b648cb4716800b0fcb3f3a0a8acb02 Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Mon, 2 Dec 2024 15:11:55 +0800
Subject: [PATCH 86/87] mtk: mac80211: relax the DFS restrictions for scanning

If user turn on the dfs relax flag by entering the following command:
echo 1 > /sys/kernel/debug/ieee80211/phy0/scan_dfs_relax
then
1. Allow the AP to scan while operating on a DFS channel
2. Allow the MLO AP to scan the other bands when 5G band is performing
   CAC (5G band will be skipped).
3. Allow legacy AP (non-5G) to scan when 5G band (MLO AP) is performing CAC.

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 include/net/cfg80211.h | 13 +++++++-
 net/mac80211/scan.c    | 70 +++++++++++++++++++++++-------------------
 net/wireless/debugfs.c | 52 +++++++++++++++++++++++++++++++
 net/wireless/nl80211.c | 27 +++++++++++-----
 net/wireless/scan.c    |  4 +--
 net/wireless/util.c    | 13 +++++---
 6 files changed, 133 insertions(+), 46 deletions(-)

diff --git a/include/net/cfg80211.h b/include/net/cfg80211.h
index 699c3f0d..b5881a75 100644
--- a/include/net/cfg80211.h
+++ b/include/net/cfg80211.h
@@ -2759,6 +2759,8 @@ struct cfg80211_scan_6ghz_params {
  * @bssid: BSSID to scan for (most commonly, the wildcard BSSID)
  * @tsf_report_link_id: for MLO, indicates the link ID of the BSS that should be
  *      used for TSF reporting. Can be set to -1 to indicate no preference.
+ * @scan_band: the bitmap of the enum nl80211_band, indicating the band used
+ *	in this scan request.
  */
 struct cfg80211_scan_request {
 	struct cfg80211_ssid *ssids;
@@ -2788,6 +2790,7 @@ struct cfg80211_scan_request {
 	u32 n_6ghz_params;
 	struct cfg80211_scan_6ghz_params *scan_6ghz_params;
 	s8 tsf_report_link_id;
+	u8 scan_band;
 
 	/* keep last */
 	struct ieee80211_channel *channels[] __counted_by(n_channels);
@@ -5766,6 +5769,8 @@ struct wiphy_radio {
  *
  * @radio: radios belonging to this wiphy
  * @n_radio: number of radios
+ *
+ * @dfs_relax: a flag to relax the DFS restrictions during scanning
  */
 struct wiphy {
 	struct mutex mtx;
@@ -5922,6 +5927,8 @@ struct wiphy {
 	int n_radio;
 	const struct wiphy_radio *radio;
 
+	bool dfs_relax;
+
 	char priv[] __aligned(NETDEV_ALIGN);
 };
 
@@ -6651,11 +6658,15 @@ bool cfg80211_radio_chandef_valid(const struct wiphy_radio *radio,
  *
  * @wdev: the wireless device
  * @chan: channel to check
+ * @radio_mask: check the channel under a user-specified radio mask.
+ *	If the radio_mask is 0, then wdev->radio_mask is used
+ *	to check the channel.
  *
  * Return: whether or not the wdev may use the channel
  */
 bool cfg80211_wdev_channel_allowed(struct wireless_dev *wdev,
-				   struct ieee80211_channel *chan);
+				   struct ieee80211_channel *chan,
+				   u32 radio_mask);
 
 /**
  * ieee80211_get_response_rate - get basic rate for a given rate
diff --git a/net/mac80211/scan.c b/net/mac80211/scan.c
index 22663f71..c47f588d 100644
--- a/net/mac80211/scan.c
+++ b/net/mac80211/scan.c
@@ -371,10 +371,9 @@ static bool ieee80211_prep_hw_scan(struct ieee80211_sub_if_data *sdata)
 	if (ieee80211_hw_check(&local->hw, SINGLE_SCAN_ON_ALL_BANDS)) {
 		local->hw_scan_req->req.n_channels = req->n_channels;
 
-		for (i = 0; i < req->n_channels; i++) {
+		for (i = 0; i < req->n_channels; i++)
 			local->hw_scan_req->req.channels[i] = req->channels[i];
-			bands_used |= BIT(req->channels[i]->band);
-		}
+		bands_used = req->scan_band;
 	} else {
 		do {
 			if (local->hw_scan_band == NUM_NL80211_BANDS)
@@ -384,8 +383,8 @@ static bool ieee80211_prep_hw_scan(struct ieee80211_sub_if_data *sdata)
 			*n_chans = 0;
 
 			for (i = 0; i < req->n_channels; i++) {
-				if (req->channels[i]->band !=
-				    local->hw_scan_band)
+				if (req->channels[i]->band != local->hw_scan_band ||
+				    !(req->scan_band & BIT(req->channels[i]->band)))
 					continue;
 				local->hw_scan_req->req.channels[(*n_chans)++] =
 							req->channels[i];
@@ -571,7 +570,8 @@ static int ieee80211_start_sw_scan(struct ieee80211_local *local,
 	return 0;
 }
 
-static bool __ieee80211_can_leave_ch(struct ieee80211_sub_if_data *sdata)
+static bool __ieee80211_can_leave_ch(struct ieee80211_sub_if_data *sdata,
+				     struct cfg80211_scan_request *req)
 {
 	struct ieee80211_local *local = sdata->local;
 	struct ieee80211_sub_if_data *sdata_iter;
@@ -582,24 +582,40 @@ static bool __ieee80211_can_leave_ch(struct ieee80211_sub_if_data *sdata)
 	if (!ieee80211_is_radar_required(local))
 		return true;
 
-	/* FIXME do not check pre-CAC allowed for scan.
-	if (!regulatory_pre_cac_allowed(local->hw.wiphy))
+	if (!local->hw.wiphy->dfs_relax &&
+	    !regulatory_pre_cac_allowed(local->hw.wiphy))
+		return false;
+
+	if (!req->scan_band)
 		return false;
-	*/
 
 	list_for_each_entry(sdata_iter, &local->interfaces, list) {
-		for_each_valid_link(&sdata_iter->wdev, link_id)
-			if (sdata_iter->wdev.links[link_id].cac_started)
-				return false;
+		for_each_valid_link(&sdata_iter->wdev, link_id) {
+			struct cfg80211_chan_def *chandef;
+
+			chandef = wdev_chandef(&sdata_iter->wdev, link_id);
+			if (!chandef || !chandef->chan ||
+			    !(req->scan_band & BIT(chandef->chan->band)))
+				continue;
+
+			if (sdata_iter->wdev.links[link_id].cac_started) {
+				if (!ieee80211_vif_is_mld(&sdata->vif))
+					return false;
+
+				/* Just disable the CAC radio for MLD */
+				req->scan_band &= ~BIT(chandef->chan->band);
+			}
+		}
 	}
 
 	return true;
 }
 
 static bool ieee80211_can_scan(struct ieee80211_local *local,
-			       struct ieee80211_sub_if_data *sdata)
+			       struct ieee80211_sub_if_data *sdata,
+			       struct cfg80211_scan_request *req)
 {
-	if (!__ieee80211_can_leave_ch(sdata))
+	if (!__ieee80211_can_leave_ch(sdata, req))
 		return false;
 
 	if (!list_empty(&local->roc_list))
@@ -622,6 +638,8 @@ void ieee80211_run_deferred_scan(struct ieee80211_local *local)
 	if (!ieee80211_can_scan(local,
 				rcu_dereference_protected(
 					local->scan_sdata,
+					lockdep_is_held(&local->hw.wiphy->mtx)),
+				rcu_dereference_protected(local->scan_req,
 					lockdep_is_held(&local->hw.wiphy->mtx))))
 		return;
 
@@ -719,10 +737,10 @@ static int __ieee80211_start_scan(struct ieee80211_sub_if_data *sdata,
 	    !(sdata->vif.active_links & BIT(req->tsf_report_link_id)))
 		return -EINVAL;
 
-	if (!__ieee80211_can_leave_ch(sdata))
+	if (!__ieee80211_can_leave_ch(sdata, req))
 		return -EBUSY;
 
-	if (!ieee80211_can_scan(local, sdata)) {
+	if (!ieee80211_can_scan(local, sdata, req)) {
 		/* wait for the work to finish/time out */
 		rcu_assign_pointer(local->scan_req, req);
 		rcu_assign_pointer(local->scan_sdata, sdata);
@@ -735,19 +753,8 @@ static int __ieee80211_start_scan(struct ieee80211_sub_if_data *sdata,
 
 		local->hw_scan_ies_bufsize = local->scan_ies_len + req->ie_len;
 
-		if (ieee80211_hw_check(&local->hw, SINGLE_SCAN_ON_ALL_BANDS)) {
-			int i, n_bands = 0;
-			u8 bands_counted = 0;
-
-			for (i = 0; i < req->n_channels; i++) {
-				if (bands_counted & BIT(req->channels[i]->band))
-					continue;
-				bands_counted |= BIT(req->channels[i]->band);
-				n_bands++;
-			}
-
-			local->hw_scan_ies_bufsize *= n_bands;
-		}
+		if (ieee80211_hw_check(&local->hw, SINGLE_SCAN_ON_ALL_BANDS))
+			local->hw_scan_ies_bufsize *= hweight8(req->scan_band);
 
 		local->hw_scan_req = kmalloc(struct_size(local->hw_scan_req,
 							 req.channels,
@@ -763,6 +770,7 @@ static int __ieee80211_start_scan(struct ieee80211_sub_if_data *sdata,
 		 * up but let UBSAN know the boundaries.
 		 */
 		local->hw_scan_req->req.n_channels = req->n_channels;
+		local->hw_scan_req->req.scan_band = req->scan_band;
 
 		ies = (u8 *)local->hw_scan_req +
 			sizeof(*local->hw_scan_req) +
@@ -1204,7 +1212,7 @@ int ieee80211_request_ibss_scan(struct ieee80211_sub_if_data *sdata,
 				if (tmp_ch->flags & (IEEE80211_CHAN_NO_IR |
 						     IEEE80211_CHAN_DISABLED) ||
 				    !cfg80211_wdev_channel_allowed(&sdata->wdev,
-								   tmp_ch))
+								   tmp_ch, 0))
 					continue;
 
 				local->int_scan_req->channels[n_ch] = tmp_ch;
@@ -1221,7 +1229,7 @@ int ieee80211_request_ibss_scan(struct ieee80211_sub_if_data *sdata,
 			if (channels[i]->flags & (IEEE80211_CHAN_NO_IR |
 						  IEEE80211_CHAN_DISABLED) ||
 			    !cfg80211_wdev_channel_allowed(&sdata->wdev,
-							   channels[i]))
+							   channels[i], 0))
 				continue;
 
 			local->int_scan_req->channels[n_ch] = channels[i];
diff --git a/net/wireless/debugfs.c b/net/wireless/debugfs.c
index a0398c7d..a1585d9c 100644
--- a/net/wireless/debugfs.c
+++ b/net/wireless/debugfs.c
@@ -407,6 +407,57 @@ dfs_available_reset(void *data, u64 val)
 DEFINE_DEBUGFS_ATTRIBUTE(dfs_available_reset_ops, NULL,
 			 dfs_available_reset, "0x%08llx\n");
 
+static ssize_t scan_dfs_relax_write(struct file *file,
+				    const char __user *user_buf,
+				    size_t count, loff_t *ppos)
+{
+	struct wiphy *wiphy = file->private_data;
+	char buf[16];
+
+	if (count >= sizeof(buf))
+		return -EINVAL;
+
+	if (copy_from_user(buf, user_buf, count))
+		return -EFAULT;
+
+	if (count && buf[count - 1] == '\n')
+		buf[count - 1] = '\0';
+	else
+		buf[count] = '\0';
+
+	if (kstrtobool(buf, &wiphy->dfs_relax))
+		return -EINVAL;
+
+	return count;
+}
+
+static ssize_t scan_dfs_relax_read(struct file *file, char __user *user_buf,
+				   size_t count, loff_t *ppos)
+{
+	struct wiphy *wiphy = file->private_data;
+	unsigned int r, offset, buf_size = PAGE_SIZE;
+	char *buf;
+
+	buf = kzalloc(buf_size, GFP_KERNEL);
+	if (!buf)
+		return -ENOMEM;
+
+	offset = scnprintf(buf, buf_size, "dfs relax: %u\n", wiphy->dfs_relax);
+
+	r = simple_read_from_buffer(user_buf, count, ppos, buf, offset);
+
+	kfree(buf);
+
+	return r;
+}
+
+static const struct file_operations scan_dfs_relax_ops = {
+	.write = scan_dfs_relax_write,
+	.read = scan_dfs_relax_read,
+	.open = simple_open,
+	.llseek = default_llseek,
+};
+
 #define DEBUGFS_ADD(name, chmod)						\
 	debugfs_create_file(#name, chmod, phyd, &rdev->wiphy, &name## _ops)
 
@@ -423,6 +474,7 @@ void cfg80211_debugfs_rdev_add(struct cfg80211_registered_device *rdev)
 	DEBUGFS_ADD(dfs_skip_nop, 0600);
 	DEBUGFS_ADD(dfs_skip_cac, 0600);
 	DEBUGFS_ADD(dfs_available_reset, 0600);
+	DEBUGFS_ADD(scan_dfs_relax, 0644);
 }
 
 struct debugfs_read_work {
diff --git a/net/wireless/nl80211.c b/net/wireless/nl80211.c
index 4b5fa813..5a266d35 100644
--- a/net/wireless/nl80211.c
+++ b/net/wireless/nl80211.c
@@ -9366,19 +9366,23 @@ int nl80211_parse_random_mac(struct nlattr **attrs,
 }
 
 static bool cfg80211_off_channel_oper_allowed(struct wireless_dev *wdev,
-					      struct ieee80211_channel *chan)
+					      struct ieee80211_channel *chan,
+					      bool dfs_relax)
 {
 	unsigned int link_id;
 	bool all_ok = true;
 
 	lockdep_assert_wiphy(wdev->wiphy);
 
-	if (!cfg80211_wdev_channel_allowed(wdev, chan))
+	if (!cfg80211_wdev_channel_allowed(wdev, chan, 0))
 		return false;
 
 	if (!cfg80211_beaconing_iface_active(wdev))
 		return true;
 
+	if (dfs_relax)
+		return true;
+
 	/*
 	 * FIXME: check if we have a free HW resource/link for chan
 	 *
@@ -9595,7 +9599,7 @@ static int nl80211_trigger_scan(struct sk_buff *skb, struct genl_info *info)
 
 			/* ignore disabled channels */
 			if (chan->flags & IEEE80211_CHAN_DISABLED ||
-			    !cfg80211_wdev_channel_allowed(wdev, chan))
+			    !cfg80211_wdev_channel_allowed(wdev, chan, 0))
 				continue;
 
 			request->channels[i] = chan;
@@ -9618,6 +9622,11 @@ static int nl80211_trigger_scan(struct sk_buff *skb, struct genl_info *info)
 		}
 	} else {
 		enum nl80211_band band;
+		u32 radio_mask;
+
+		err = nl80211_parse_vif_radio_mask(info, &radio_mask);
+		if (err < 0)
+			goto out_free;
 
 		/* all channels */
 		for (band = 0; band < NUM_NL80211_BANDS; band++) {
@@ -9631,7 +9640,7 @@ static int nl80211_trigger_scan(struct sk_buff *skb, struct genl_info *info)
 				chan = &wiphy->bands[band]->channels[j];
 
 				if (chan->flags & IEEE80211_CHAN_DISABLED ||
-				    !cfg80211_wdev_channel_allowed(wdev, chan))
+				    !cfg80211_wdev_channel_allowed(wdev, chan, radio_mask))
 					continue;
 
 				request->channels[i] = chan;
@@ -9650,8 +9659,9 @@ static int nl80211_trigger_scan(struct sk_buff *skb, struct genl_info *info)
 	for (i = 0; i < request->n_channels; i++) {
 		struct ieee80211_channel *chan = request->channels[i];
 
+		request->scan_band |= BIT(chan->band);
 		/* if we can go off-channel to the target channel we're good */
-		if (cfg80211_off_channel_oper_allowed(wdev, chan))
+		if (cfg80211_off_channel_oper_allowed(wdev, chan, wiphy->dfs_relax))
 			continue;
 
 		if (!cfg80211_wdev_on_sub_chan(wdev, chan, true)) {
@@ -12826,7 +12836,10 @@ static int nl80211_remain_on_channel(struct sk_buff *skb,
 	if (err)
 		return err;
 
-	if (!cfg80211_off_channel_oper_allowed(wdev, chandef.chan)) {
+	if (wdev->links[link_id].cac_started)
+		return -EBUSY;
+
+	if (!cfg80211_off_channel_oper_allowed(wdev, chandef.chan, rdev->wiphy.dfs_relax)) {
 		const struct cfg80211_chan_def *oper_chandef, *compat_chandef;
 
 		oper_chandef = wdev_chandef(wdev, link_id);
@@ -13043,7 +13056,7 @@ static int nl80211_tx_mgmt(struct sk_buff *skb, struct genl_info *info)
 		return -EINVAL;
 
 	if (params.offchan &&
-	    !cfg80211_off_channel_oper_allowed(wdev, chandef.chan))
+	    !cfg80211_off_channel_oper_allowed(wdev, chandef.chan, false))
 		return -EBUSY;
 
 	params.link_id = nl80211_link_id_or_invalid(info->attrs);
diff --git a/net/wireless/scan.c b/net/wireless/scan.c
index e0f36a9d..090dba76 100644
--- a/net/wireless/scan.c
+++ b/net/wireless/scan.c
@@ -957,7 +957,7 @@ static int cfg80211_scan_6ghz(struct cfg80211_registered_device *rdev)
 			ieee80211_get_channel(&rdev->wiphy, ap->center_freq);
 
 		if (!chan || chan->flags & IEEE80211_CHAN_DISABLED ||
-		    !cfg80211_wdev_channel_allowed(rdev_req->wdev, chan))
+		    !cfg80211_wdev_channel_allowed(rdev_req->wdev, chan, 0))
 			continue;
 
 		for (i = 0; i < rdev_req->n_channels; i++) {
@@ -3520,7 +3520,7 @@ int cfg80211_wext_siwscan(struct net_device *dev,
 			/* ignore disabled channels */
 			chan = &wiphy->bands[band]->channels[j];
 			if (chan->flags & IEEE80211_CHAN_DISABLED ||
-			    !cfg80211_wdev_channel_allowed(creq->wdev, chan))
+			    !cfg80211_wdev_channel_allowed(creq->wdev, chan, 0))
 				continue;
 
 			/* If we have a wireless request structure and the
diff --git a/net/wireless/util.c b/net/wireless/util.c
index c03b185d..0feeb92f 100644
--- a/net/wireless/util.c
+++ b/net/wireless/util.c
@@ -2933,21 +2933,24 @@ bool cfg80211_radio_chandef_valid(const struct wiphy_radio *radio,
 EXPORT_SYMBOL(cfg80211_radio_chandef_valid);
 
 bool cfg80211_wdev_channel_allowed(struct wireless_dev *wdev,
-				   struct ieee80211_channel *chan)
+				   struct ieee80211_channel *chan,
+				   u32 radio_mask)
 {
 	struct wiphy *wiphy = wdev->wiphy;
 	const struct wiphy_radio *radio;
 	struct cfg80211_chan_def chandef;
-	u32 radio_mask;
+	u32 mask = radio_mask & wdev->radio_mask;
 	int i;
 
-	radio_mask = wdev->radio_mask;
-	if (!wiphy->n_radio || radio_mask == BIT(wiphy->n_radio) - 1)
+	if (!radio_mask)
+		mask = wdev->radio_mask;
+
+	if (!wiphy->n_radio || mask == BIT(wiphy->n_radio) - 1)
 		return true;
 
 	cfg80211_chandef_create(&chandef, chan, NL80211_CHAN_HT20);
 	for (i = 0; i < wiphy->n_radio; i++) {
-		if (!(radio_mask & BIT(i)))
+		if (!(mask & BIT(i)))
 			continue;
 
 		radio = &wiphy->radio[i];
-- 
2.45.2

